diff --git a/hal/base_example/src/main.c b/hal/base_example/src/main.c
index d0e5655..a276e2f 100644
--- a/hal/base_example/src/main.c
+++ b/hal/base_example/src/main.c
@@ -1,4 +1,6 @@
 #include "main.h"
+#include <stdio.h>
+#include <string.h>
 
 ADC_HandleTypeDef hadc1;
 ADC_HandleTypeDef hadc2;
@@ -10,6 +12,14 @@ RTC_HandleTypeDef hrtc;
 
 PCD_HandleTypeDef hpcd_USB_OTG_FS;
 
+#define ADC_BUF_LEN 4096
+
+uint32_t adc_buf[ADC_BUF_LEN] = {0};
+char serial_buffer[64] = {0};
+
+uint16_t adcResult = 0;
+float adcResultFloatinMV = 0;
+
 void SystemClock_Config(void);
 void PeriphCommonClock_Config(void);
 static void MX_GPIO_Init(void);
@@ -20,6 +30,18 @@ static void MX_LPUART1_UART_Init(void);
 static void MX_USB_OTG_FS_PCD_Init(void);
 static void MX_RTC_Init(void);
 
+void HAL_ADC_ConvHalfCpltCallback(ADC_HandleTypeDef *hadc) {
+    if (hadc == &hadc2) {
+        HAL_GPIO_TogglePin(GPIOB, LD3_Pin);
+    }
+}
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *hadc) {
+    if (hadc == &hadc2) {
+        HAL_GPIO_TogglePin(GPIOB, LD2_Pin);
+    }
+}
+
 int main(void) {
 
     HAL_Init();
@@ -36,7 +58,18 @@ int main(void) {
     MX_USB_OTG_FS_PCD_Init();
     MX_RTC_Init();
 
+    HAL_ADC_Start_DMA(&hadc2, adc_buf, ADC_BUF_LEN);
+
     while (1) {
+        HAL_ADC_Start(&hadc1);
+        HAL_ADC_PollForConversion(&hadc1, HAL_MAX_DELAY);
+        adcResult = HAL_ADC_GetValue(&hadc1);
+        adcResultFloatinMV = ((3300 * adcResult) / 4095);
+        memset(serial_buffer, 0, sizeof(serial_buffer));
+        snprintf(serial_buffer, sizeof(serial_buffer), "MV: %d\n", (int)adcResultFloatinMV);
+        HAL_UART_Transmit(&hlpuart1, (uint8_t *)serial_buffer, strlen(serial_buffer), 100);
+
+        HAL_Delay(1000);
     }
 }
 
